toolchain/lowering/lowering_context.cpp

```diff
@@ -47,19 +48,56 @@ auto LoweringContext::LowerBlock(SemanticsNodeBlockId block_id) -> void {
   }
 }
 
-auto LoweringContext::LowerNodeToType(SemanticsNodeId node_id) -> llvm::Type* {
-  CARBON_CHECK(node_id.is_valid());
+auto LoweringContext::BuildLoweredNodeAsType(SemanticsNodeId node_id)
+    -> llvm::Type* {
   switch (node_id.index) {
+    case SemanticsBuiltinKind::EmptyStructType.AsInt():
     case SemanticsBuiltinKind::EmptyTuple.AsInt():
-      // TODO: Should probably switch this to an actual empty tuple in the
-      // future, but it's implemented as void for now.
-      return builder_.getVoidTy();
+    case SemanticsBuiltinKind::EmptyTupleType.AsInt():
+      // Represent empty data as 0-length arrays.
+      // TODO: Work to remove EmptyTuple here.
+      return llvm::ArrayType::get(builder_.getInt8Ty(), 0);
```

Comment:
Does the [void type] not work here (maybe because you need to form values of it)?

If not, I'd suggest an empty struct type rather than an empty array.

[void type]: https://llvm.org/docs/LangRef.html#void-type

Reply:
Yes, specifically `var x: {} = {};` is not valid if `{}` is the void type.

Reply:
I should note, `var x: {};` is the actual crash, you can't have void there.

Reply:
Almost forgot -- switched to the empty struct type.